libsolutil/IpfsHash.cpp

```diff
@@ -53,36 +53,127 @@ string base58Encode(bytes const& _data)
 	reverse(output.begin(), output.end());
 	return output;
 }
+struct Chunk
+{
+	Chunk() = default;
+	Chunk(bytes _hash, size_t _size, size_t _blockSize):
+		hash(std::move(_hash)),
+		size(_size),
+		blockSize(_blockSize)
+	{}
+	bytes hash;
+	size_t size;
+	size_t blockSize;
+};
+
+using Chunks = vector<Chunk>;
+
+/// Builds a tree starting from the bottom level where nodes are data nodes.
+/// Data nodes should be calculated and passed as the only level in chunk levels
+/// Each next level is calculated as following:
+///   - Pick up to maxChildNum (174) nodes until a whole level is added, group them and pass to the node in the next level
+///   - Do this until the current level has only one node, return the hash in that node
+bytes groupChunksBottomUp(vector<Chunks>& chunkLevels)
+{
+	size_t const maxChildNum = 174;
+	size_t currentLevel = 0;
+
+	// when we reach root it will be the only node in that level
+	while (chunkLevels[currentLevel].size() != 1)
+	{
+		vector<Chunk> nextLevel;
+		Chunk newChunk;
+		bytes newLinks;
+		bytes newLengths;
+		size_t newChunkSize = 0;
+		size_t newBlockSize = 0;
+
+		for (size_t chunkIndex = 0; chunkIndex < chunkLevels[currentLevel].size(); chunkIndex++)
```

Comment:
This loop only uses `chunkLevels` as `chunkLevels[currentLevel]` - is it possible to extract it into its own function?